sw/device/silicon_creator/lib/base/sec_mmio.c

```diff
@@ -123,51 +111,39 @@ void sec_mmio_write_increment(uint32_t value) {
 }
 
 void sec_mmio_check_values(uint32_t rnd_offset) {
-  size_t offset = rnd_offset;
+  // Pick a random starting offset.
+  uint32_t offset =
+      ((uint64_t)rnd_offset * (uint64_t)sec_mmio_ctx.last_index) >> 32;
+  uint32_t step = 1;
   size_t i;
-  for (i = 0; i < sec_mmio_ctx.last_index; ++i) {
-    // FIXME: Remove dependency on __udivdi3.
-    offset = (offset + kSecMmioRndStep) % sec_mmio_ctx.last_index;
+  for (i = 0; launder32(i) < sec_mmio_ctx.last_index; ++i) {
     uint32_t read_value = abs_mmio_read32(sec_mmio_ctx.addrs[offset]);
-    if ((read_value ^ kSecMmioMaskVal) != sec_mmio_ctx.values[offset]) {
-      sec_mmio_shutdown_cb(kErrorSecMmioCheckValueFault);
-      __builtin_unreachable();
+    HARDENED_CHECK_EQ((read_value ^ kSecMmioMaskVal),
+                      sec_mmio_ctx.values[offset]);
+    offset += step;
+    if (offset >= sec_mmio_ctx.last_index) {
+      offset -= sec_mmio_ctx.last_index;
     }
   }
   // Check for loop completion.
-  if (i != sec_mmio_ctx.last_index) {
-    sec_mmio_shutdown_cb(kErrorSecMmioCheckIndexFault);
-    __builtin_unreachable();
-  }
+  HARDENED_CHECK_EQ(launder32(i), sec_mmio_ctx.last_index);
   ++sec_mmio_ctx.check_count;
 }
 
 void sec_mmio_check_counters(uint32_t expected_check_count) {
-  // Generous use of volatile in critical variables to avoid compiler
-  // optimizations, and map "zero" to a value with good hamming weight
-  // and with a good hamming distance to "all-ones".
-  // TODO(#6610): Update based on implementation guidance.
-  static volatile const uint32_t kValZero = 0x3ca5965a;
-  static volatile const uint32_t kValOnes = 0xc35a69a5;
-
-  uint32_t result = kValZero ^ sec_mmio_ctx.write_count;
-  result ^= sec_mmio_ctx.expected_write_count;
+  uint32_t result = kSecMmioValZero ^ sec_mmio_ctx.write_count;
+  result = launder32(result) ^ sec_mmio_ctx.expected_write_count;
 
   // Check the expected write count. This is equivalent to
   // sec_mmio_ctx.write_count == sec_mmio_ctx.expected_write_count
-  if (result != kValZero) {
-    sec_mmio_shutdown_cb(kErrorSecMmioWriteCountFault);
-    __builtin_unreachable();
-  }
+  HARDENED_CHECK_EQ(launder32(result), kSecMmioValZero);
 
   // Check the expected check counts. This is equivalent to
   // sec_mmio_ctx.check_count == expected_check_count. This check is expected to
   // fail if the previous check failed.
-  result ^= sec_mmio_ctx.check_count;
-  result ^= expected_check_count;
-  if (~result != kValOnes) {
-    sec_mmio_shutdown_cb(kErrorSecMmioCheckCountFault);
-    __builtin_unreachable();
-  }
+  result = launder32(result) ^ sec_mmio_ctx.check_count;
+  result = launder32(result) ^ expected_check_count;
+  HARDENED_CHECK_EQ(launder32(~result), kSecMmioValOne);
```

Comment:
This may be worth writing as `~launder32(result)`, because otherwise the compiler can rewrite this and expand it to
```c
launder32(~result) -> launder32(launder32(result) ^ ~expected_check_count);
```
Right now this probably won't have an adverse optimization result, but better safe than sorry?

Reply:
Ended up using `HARDENED_CHECK_EQ(~launder32(result), kSecMmioValOne)` over `HARDENED_CHECK_EQ(~result, kSecMmioValOne)`.